src/charm.py

```diff
@@ -120,21 +121,30 @@ def _on_provide_certificate_action(self, event: ActionEvent) -> None:
         certificate = base64.b64decode(event.params["certificate"]).decode("utf-8").strip()
         ca_cert = base64.b64decode(event.params["ca-certificate"]).decode("utf-8").strip()
 
-        if not self._csr_exists_in_requirer(csr=csr, relation_id=event.params["relation-id"]):
-            event.fail(message="Certificate signing request was not found in requirer data.")
-            return
-
         if not csr_matches_certificate(csr=csr, cert=certificate):
             event.fail(message="Certificate and CSR do not match.")
             return
 
+        found_relation_ids = []
+        for relation in self.model.relations.get("certificates", []):
+            if self._csr_exists_in_requirer(csr=csr, relation_id=relation.id):
+                found_relation_ids.append(relation.id)
+
+        requested_relation_id = event.params.get("relation-id", None)
+        err = self._relation_id_parameter_valid(found_relation_ids, requested_relation_id)
+        if err:
+            event.fail(message=err)
+            return
+
         try:
             self.tls_certificates.set_relation_certificate(
                 certificate_signing_request=csr,
                 certificate=certificate,
                 ca=ca_cert,
                 chain=ca_chain_list,
-                relation_id=event.params["relation-id"],
+                relation_id=(
+                    requested_relation_id if requested_relation_id else found_relation_ids[0]
```

Comment:
Is there any use case possible where 2 requirers send the same csr?
I don't think so, but then maybe we could make it explicit and not use a list for `found_relation_ids`?

Reply:
It is possible, and we catch it here. See the discussion [here](https://matrix.to/#/!yAkGlrYcBFYzYRvOlQ%3Aubuntu.com/%24ZsnWJqY_MZp-E3qIoDH0Q6x8rnUAIywJ6WSB7KKcqx4).

Reply:
What do you mean by catch i?

Reply:
https://github.com/canonical/manual-tls-certificates-operator/blob/4dc286ecc553256bc010539d997619ee813c7d55/src/charm.py#L185

Reply:
Changed the docstring to clarify this a bit more

Reply:
Gotcha, thanks :)